stan/math/prim/mat/fun/mdivide_left_tri.hpp

```diff
@@ -52,6 +58,77 @@ inline Eigen::Matrix<T, R1, C1> mdivide_left_tri(
   return b;
 }
 
+/**
+ * Returns the solution of the system Ax=b when A is triangular
+ * and A and b are matrices of doubles.
+ * @param A Triangular matrix.  Specify upper or lower with TriView
+ * being Eigen::Upper or Eigen::Lower.
+ * @param b Right hand side matrix or vector.
+ * @return x = A^-1 b, solution of the linear system.
+ * @throws std::domain_error if A is not square or the rows of b don't
+ * match the size of A.
+ */
+template <int TriView, int R1, int C1, int R2, int C2>
+inline Eigen::Matrix<double, R1, C2> mdivide_left_tri(
+    const Eigen::Matrix<double, R1, C1> &A,
+    const Eigen::Matrix<double, R2, C2> &b) {
+  check_square("mdivide_left_tri", "A", A);
+  check_multiplicable("mdivide_left_tri", "A", A, "b", b);
+#ifdef STAN_OPENCL
+  if (A.rows()
+      >= opencl_context.tuning_opts().lower_tri_inverse_size_worth_transfer) {
+    matrix_cl A_cl(A);
+    matrix_cl b_cl(b);
+    matrix_cl A_inv_cl(A.rows(), A.cols());
+    if (TriView == Eigen::Lower) {
+      A_inv_cl = lower_triangular_inverse(A_cl);
+    } else {
+      A_inv_cl = transpose(lower_triangular_inverse(transpose(A_cl)));
+    }
+    matrix_cl C_cl = A_inv_cl * b_cl;
+    return from_matrix_cl(C_cl);
+  } else {
+#endif
+    return A.template triangularView<TriView>().solve(b);
+#ifdef STAN_OPENCL
+  }
+#endif
+}
+
+/**
+ * Returns the solution of the system Ax=b when A is triangular, b=I and
+ * both are matrices of doubles.
+ * @param A Triangular matrix.  Specify upper or lower with TriView
+ * being Eigen::Upper or Eigen::Lower.
+ * @return x = A^-1 .
+ * @throws std::domain_error if A is not square
+ */
+template <int TriView, int R1, int C1>
+inline Eigen::Matrix<double, R1, C1> mdivide_left_tri(
+    const Eigen::Matrix<double, R1, C1> &A) {
+  check_square("mdivide_left_tri", "A", A);
+  const int n = A.rows();
+#ifdef STAN_OPENCL
+  if (A.rows()
+      >= opencl_context.tuning_opts().lower_tri_inverse_size_worth_transfer) {
+    matrix_cl A_cl(A);
+    if (TriView == Eigen::Lower) {
+      A_cl = lower_triangular_inverse(A_cl);
```

Comment:
Since `A_cl` is used as input/output here maybe we should have one `lower_triangular_inverse` that returns back a matrix and another that is a method in `matrix_cl` that operates on `*this`

Reply:
What would be the best solution to not duplicate all the `tri_inverse` code? The inplace one being the one with all the existing tri_inverse code and the regular option just call the inplace one?
So something like this: 
```cpp
template <TriangularViewCL triangular_view>
inline void matrix_cl::tri_inverse() try {
//existing code on *this
}

template <TriangularViewCL triangular_view>
inline matrix_cl tri_inverse(const matrix_cl& A) {
matrix_cl B(A);
B.tri_inverse<triangular_view>();
return B;
}
```

Reply:
Also not really sure we want to do this, since all matrix_cl functions are actually inplace, with inplace kernels. tri_inverse isnt

Reply:
Ah, crap yeah sorry I got confused by the above assinging to it's input my b . Yeah this returns a new matrix so that's fine as is